File: src/main/java/io/github/xf8b/xf8bot/commands/info/HelpCommand.java
/*
 * Copyright (c) 2020 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.commands;

import com.google.common.collect.Range;
import discord4j.core.spec.EmbedCreateSpec;
import discord4j.rest.util.Color;
import discord4j.rest.util.Permission;
import io.github.xf8b.xf8bot.Xf8bot;
import io.github.xf8b.xf8bot.api.commands.AbstractCommand;
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent;
import io.github.xf8b.xf8bot.api.commands.CommandRegistry;
import io.github.xf8b.xf8bot.api.commands.arguments.IntegerArgument;
import io.github.xf8b.xf8bot.api.commands.arguments.StringArgument;
import org.apache.commons.text.WordUtils;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public class HelpCommand extends AbstractCommand {
    //TODO: make paginated embed system so this can go back to using reactions for pages
    //public static Message currentMessage = null;
    //public static CommandType currentCommandType = null;
    //public static final List<AbstractCommand> commandsShown = Collections.synchronizedList(new ArrayList<>());

    private static final StringArgument SECTION_OR_COMMAND = StringArgument.builder()
            .setIndex(Range.singleton(1))
            .setName("section or command")
            .setNotRequired()
            .build();
    private static final IntegerArgument PAGE = IntegerArgument.builder()
            .setIndex(Range.singleton(2))
            .setName("page")
            .setNotRequired()
            .build();

    public HelpCommand() {
        super(AbstractCommand.builder()
                .setName("${prefix}help")
                .setDescription("""
                        Shows the command's description, usage, aliases, and actions.
                        If no command was specified, all the commands in the section will be shown.
                        If no section was specified, all the commands will be shown.
                        """)
                .setCommandType(CommandType.INFO)
                .setArguments(SECTION_OR_COMMAND, PAGE)
                .setBotRequiredPermissions(Permission.EMBED_LINKS));
    }

    @NotNull
    @Override
    public Mono<Void> onCommandFired(@NotNull CommandFiredEvent event) {
        String guildId = event.getGuild().block().getId().asString();
        Xf8bot xf8bot = event.getXf8bot();
        Optional<String> commandOrSection = event.getValueOfArgument(SECTION_OR_COMMAND);
        if (commandOrSection.isEmpty()) {
            return event.getChannel().flatMap(it -> it.createEmbed(embedCreateSpec -> {
                embedCreateSpec.setTitle("Help Page")
                        .setColor(Color.BLUE);
                for (CommandType commandType : CommandType.values()) {
                    String commandTypeName = WordUtils.capitalizeFully(commandType.name()
                            .toLowerCase()
                            .replace("_", " "));
                    embedCreateSpec.addField(
                            "`" + commandTypeName + "`",
                            commandType.getDescription() + "\n" +
                                    "To go to this section, use `" + event.getPrefix().block() + "help " + commandType.name()
                                    .toLowerCase()
                                    .replace(" ", "_") + "`.",
                            false
                    );
                }
            })).then();
        } else {
            for (CommandType commandType : CommandType.values()) {
                if (commandOrSection.get().equalsIgnoreCase(commandType.name())) {
                    Integer pageNumber = event.getValueOfArgument(PAGE).orElse(0);
                    List<AbstractCommand> commandsWithCurrentCommandType = event.getXf8bot().getCommandRegistry()
                            .getCommandsWithCommandType(commandType);
                    if (commandsWithCurrentCommandType.size() > 6) {
                        if (!Range.closedOpen(0, commandsWithCurrentCommandType.size() % 6).contains(pageNumber)) {
                            return event.getChannel()
                                    .flatMap(it -> it.createMessage("No page with the index " + (pageNumber + 1) + " exists!"))
                                    .then();
                        }
                    }
                    //commandsShown.clear();
                    //currentCommandType = commandType;
                    int finalPageNumber = pageNumber;
                    return event.getChannel().flatMap(it -> it.createEmbed(embedCreateSpec -> generateCommandTypeEmbed(
                            event,
                            event.getXf8bot().getCommandRegistry(),
                            embedCreateSpec,
                            commandType,
                            guildId,
                            finalPageNumber
                    ))).then();//.subscribe(/*message -> {
                    //message.addReaction(ReactionEmoji.unicode("⬅")).block();
                    //message.addReaction(ReactionEmoji.unicode("➡")).block();
                    //currentMessage = message;
                    //}*/);
                }
            }
            for (AbstractCommand command : event.getXf8bot().getCommandRegistry()) {
                String name = command.getName();
                String nameWithPrefix = command.getNameWithPrefix(xf8bot, guildId);
                List<String> aliases = command.getAliases();
                List<String> aliasesWithPrefixes = command.getAliasesWithPrefixes(xf8bot, guildId);
                if (commandOrSection.get().equals(name.replace("${prefix}", ""))) {
                    String description = command.getDescription();
                    String usage = command.getUsageWithPrefix(xf8bot, guildId);
                    Map<String, String> actions = command.getActions();
                    return event.getChannel().flatMap(it -> it.createEmbed(embedCreateSpec -> generateCommandEmbed(
                            embedCreateSpec,
                            nameWithPrefix,
                            description,
                            usage,
                            aliasesWithPrefixes,
                            actions
                    ))).then();
                } else if (!aliases.isEmpty()) {
                    for (String alias : aliases) {
                        if (commandOrSection.get().equals(alias.replace("${prefix}", ""))) {
                            String description = command.getDescription();
                            String usage = command.getUsageWithPrefix(xf8bot, guildId);
                            Map<String, String> actions = command.getActions();
                            return event.getChannel().flatMap(it -> it.createEmbed(embedCreateSpec -> generateCommandEmbed(
                                    embedCreateSpec,
                                    nameWithPrefix,
                                    description,
                                    usage,
                                    aliasesWithPrefixes,
                                    actions
                            ))).then();
                        }
                    }
                }
            }
        }
        return Mono.empty(); //TODO: see if this breaks anything
    }

    private void generateCommandTypeEmbed(@NotNull CommandFiredEvent event, @NotNull CommandRegistry commandRegistry, @NotNull EmbedCreateSpec embedCreateSpec, CommandType commandType, @NotNull String guildId, int pageNumber) {
        Xf8bot xf8bot = event.getXf8bot();
        embedCreateSpec.setTitle("Help Page #" + (pageNumber + 1))
                .setDescription("Actions are not listed on this page. To see them, do `" + event.getPrefix().block() + "help <command>`.\n" +
                        "To go to a different page, use `" + event.getPrefix().block() + "help <section> <page>`.")
                .setColor(Color.BLUE);
        List<AbstractCommand> commandsWithCurrentCommandType = commandRegistry
                .getCommandsWithCommandType(commandType);
        for (int i = pageNumber * 6; i < pageNumber * 6 + 6; i++) {
            AbstractCommand command;
            try {
                command = commandsWithCurrentCommandType.get(i);
            } catch (IndexOutOfBoundsException exception) {
                break;
            }
            String name = command.getNameWithPrefix(xf8bot, guildId);
            String nameWithPrefixRemoved = command.getName().replace("${prefix}", "");
            String description = command.getDescription();
            String usage = command.getUsageWithPrefix(xf8bot, guildId);
            embedCreateSpec.addField(
                    "`" + name + "`",
                    description + "\n" +
                            "Usage: `" + usage + "`\n" +
                            "If you want to go to the help page for this command, use `" + event.getPrefix().block() + "help " + nameWithPrefixRemoved + "`.",
                    false
            );
        }
    }

    private void generateCommandEmbed(@NotNull EmbedCreateSpec embedCreateSpec, String name, String description, String usage, @NotNull List<String> aliases, @NotNull Map<String, String> actions) {
        embedCreateSpec.setTitle("Help Page For `" + name + "`")
                .addField("`" + name + "`", description + "\n" +
                        "Usage: `" + usage + "`", false)
                .setColor(Color.BLUE);
        if (!actions.isEmpty()) {
            StringBuilder actionsFormatted = new StringBuilder();
            actions.forEach((action, actionDescription) -> actionsFormatted.append("`").append(action).append("`: ").append(actionDescription).append("\n"));
            embedCreateSpec.addField("Actions", actionsFormatted.toString().replaceAll("\n$", ""), false);
        }
        if (!aliases.isEmpty()) {
            StringBuilder aliasesFormatted = new StringBuilder();
            aliases.forEach(alias -> aliasesFormatted.append("`").append(alias).append("`\n"));
            embedCreateSpec.addField("Aliases", aliasesFormatted.toString().replaceAll("\n$", ""), false);
        }
    }
}

File: src/main/kotlin/io/github/xf8b/xf8bot/commands/info/HelpCommand.kt
/*
 * Copyright (c) 2020, 2021 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.commands.info

import com.google.common.collect.Range
import discord4j.common.util.Snowflake
import discord4j.rest.util.Color
import discord4j.rest.util.Permission
import io.github.xf8b.xf8bot.api.commands.Command
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent
import io.github.xf8b.xf8bot.api.commands.arguments.StringArgument
import io.github.xf8b.xf8bot.util.EmbedCreateDsl
import io.github.xf8b.xf8bot.util.createEmbedDsl
import io.github.xf8b.xf8bot.util.extensions.*
import io.github.xf8b.xf8bot.util.immutableListOf
import io.github.xf8b.xf8bot.util.pagination.createPaginatedEmbed
import org.apache.commons.text.WordUtils
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.toFlux
import reactor.kotlin.core.publisher.toMono
import java.util.*

class HelpCommand : Command(
    name = "\${prefix}help",
    description = """
    If a command was specified, this shows the command's description, usage, aliases, and actions.
    If no command was specified, but a section was specified, all the commands in the section will be shown.
    If no section or command was specified, all the commands will be shown.
    """.trimIndent(),
    commandType = CommandType.INFO,
    arguments = immutableListOf(SECTION_OR_COMMAND),
    botRequiredPermissions = Permission.EMBED_LINKS.toSingletonPermissionSet()
) {
    override fun onCommandFired(event: CommandFiredEvent): Mono<Void> {
        val commandOrSection = event[SECTION_OR_COMMAND]

        if (commandOrSection == null) {
            return event.prefix
                .map { prefix -> if (prefix.isAlpha()) "$prefix " else prefix }
                .flatMap { prefix ->
                    event.channel.flatMap {
                        it.createEmbedDsl {
                            title("Help Page")
                            color(Color.BLUE)

                            for (commandType in CommandType.values()) {
                                val commandTypeName = commandType.name
                                    .toLowerCase(Locale.ROOT)
                                    .replace("_", " ")

                                field(
                                    "`${WordUtils.capitalizeFully(commandTypeName)}`",
                                    """
                                    ${commandType.description}
                                    To go to this section, use `${prefix}help $commandTypeName`
                                    """.trimIndent(),
                                    inline = false
                                )
                            }
                        }
                    }
                }
                .then()
        } else {
            for (commandType in CommandType.values()) {
                if (commandOrSection.equals(commandType.name, ignoreCase = true)) {
                    return generateCommandTypeEmbed(event, commandType)
                }
            }

            for (command in event.xf8bot.commandRegistry) {
                if (commandOrSection == command.rawName || command.rawAliases.any { alias -> commandOrSection == alias }) {
                    return generateCommandEmbed(event, command)
                }
            }
        }

        return event.channel
            .flatMap { it.createMessage("Could not find command/section $commandOrSection!") }
            .then()
    }

    private fun generateCommandTypeEmbed(event: CommandFiredEvent, commandType: CommandType): Mono<Void> =
        event.xf8bot.commandRegistry
            .getCommandsWithCommandType(commandType)
            .toFlux()
            .flatMap { command ->
                Mono.zip(
                    command.getNameWithPrefix(event.xf8bot, event.guildId.map(Snowflake::asString).get()),
                    command.rawName.toMono(),
                    command.description.toMono(),
                    command.getUsageWithPrefix(event.xf8bot, event.guildId.map(Snowflake::asString).get()),
                    command.administratorLevelRequired.toMono(),
                    event.prefix
                )
            }
            .buffer(6)
            .map<EmbedCreateDsl.() -> Unit> { information ->
                {
                    val prefix = information[0].t6

                    title("Help Page")
                    description("More detailed command information is not listed on this page. To see it, use `${prefix}help <command>`.")

                    for (info in information) {
                        val (name, rawName, description, usage, administratorLevel) = info

                        field(
                            "`$name`",
                            """
                            $description
                            Usage: `$usage`
                            Administrator Level Required: $administratorLevel
                            If you want to go to the help page for this command, use `${prefix}help $rawName`.
                            """.trimIndent(),
                            inline = false
                        )
                    }

                    color(Color.BLUE)
                }
            }
            .collectList()
            .flatMap { dslInitializers ->
                event.channel.flatMap { channel ->
                    channel.createPaginatedEmbed(*dslInitializers.toTypedArray())
                }
            }
            .then()

    private fun generateCommandEmbed(event: CommandFiredEvent, command: Command): Mono<Void> = Mono.zip(
        command.getUsageWithPrefix(event.xf8bot, event.guildId.map(Snowflake::asString).get()),
        command.getAliasesWithPrefixes(event.xf8bot, event.guildId.map(Snowflake::asString).get()).collectList()
    ).flatMap { commandInfo ->
        event.channel.flatMap { channel ->
            channel.createEmbedDsl {
                val (usage, aliases) = commandInfo

                title("Help Page For `${command.rawName}`")

                field(
                    "`${command.rawName}`",
                    """
                    ${command.description}
                    Usage: `$usage`
                    """.trimIndent(),
                    inline = false
                )

                field(
                    "Administrator Level Required",
                    command.administratorLevelRequired.toString(),
                    inline = false
                )
                field(
                    "Bot Required Permissions",
                    command.botRequiredPermissions
                        .joinToString { WordUtils.capitalizeFully(it.name.replace("_", " ")) }
                        .takeIf(String::isNotEmpty)
                        ?: "No permissions required",
                    inline = false
                )

                color(Color.BLUE)

                if (command.actions.isNotEmpty()) {
                    val actionsFormatted = command.actions
                        .toList()
                        .joinToString(separator = "\n") { "`${it.first}`: ${it.second}" }

                    field("Actions", actionsFormatted, inline = false)
                }

                if (aliases.isNotEmpty()) {
                    val aliasesFormatted = aliases.joinToString(separator = "\n") { "`$it`" }

                    field("Aliases", aliasesFormatted, inline = false)
                }
            }
        }
    }.then()

    companion object {
        private val SECTION_OR_COMMAND = StringArgument(
            name = "section or command",
            index = Range.singleton(0),
            required = false
        )
    }
}